jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java

```diff
@@ -434,22 +441,22 @@ private void checkAccess(Class<?> targetClass, String name, int memberModifiers,
 				/* checkClassAccess already determined that we have more than "no access" (public access) */
 				return;
 			} else if (Modifier.isPrivate(memberModifiers)) {
-				if (Modifier.isPrivate(accessMode) && ((targetClass == accessClass)
+				if (Modifier.isPrivate(accessMode) && ((definingClass == accessClass)
 /*[IF Java11]*/
-						|| targetClass.isNestmateOf(accessClass)
+						|| definingClass.isNestmateOf(accessClass)
 /*[ENDIF] Java11*/	
 				)) {
 					return;
 				}
 			} else if (Modifier.isProtected(memberModifiers)) {
 				/* Ensure that the accessMode is not restricted (public-only) */
 				if (PUBLIC != accessMode) {
-					if (targetClass.isArray()) {
+					if (definingClass.isArray()) {
 						/* The only methods array classes have are defined on Object and thus accessible */
 						return;
 					}
 					
-					if (isSamePackage(accessClass, targetClass)) {
+					if (isSamePackage(accessClass, referenceClass)) {
```

Comment:
The API Spec at https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/invoke/MethodHandles.Lookup.html
```
If the desired member is protected, the usual JVM rules apply, including the requirement that the lookup class 
must be either be in the same package as the desired member, or must inherit that member. 
(See the Java Virtual Machine Specification, sections 4.9.2, 5.4.3.5, and 6.4.) In addition, if the desired member
 is a non-static field or method in a different package, the resulting method handle may only be applied to 
 objects of the lookup class or one of its subclasses. 
```
So changing to referenceClass seems incorrect at this point.

Reply:
In this case, the method is not public, so we need to check for package access.  Consider the case where the lookup class is in the same package as the reference class, but a different package from the defining class.  If we check against the defining class we would fail, even though the method is visible via the reference class.

Note also that the reference class may broaden the visibility of the method.